### DISCUSSION TITLE
Custom Piece Development Lifecycle

### DISCUSSION THREAD

**Discussion by tdari** [Tauffer-Consulting/domino#291] [Q&A] [answered]
Hey there, recently I started to implement my custom piece implementation. The problem is I have to release and see how things work in the workflow I created, but this can be very tedious and time consuming. Whenever I change my components I have to release a new package for a simple code change and wait for the actions to push it to Github registry. Is there any local deployment option? What is your typical workflow when you're developing your pieces? Thanks.

**Comment by luiztauffer** [comment]
hi @tdari for the simple case of unit testing, you can use the `piece_dry_run` function: https://docs.domino-workflows.io/pieces/pieces_unit_testing

**Reply by tdari** [reply]
Hi @luiztauffer, yes I do unit testing what I'm looking for is a way of building the image locally when I update the piece function and let Domino build containers from this image without a need to push and pull to the registry.

**Reply by vinicvaz** [reply, marked as answer]
Hey @tdari.
Short answer, yes, it can be done, but but unfortunately we don't have a direct interface with our CLI for this yet. It be done with a small code modification in your local domino environment. I'll try to create an easy interface asap for this.

While we don't have it, you can do the following:

1. Clone Domino
2. In the root of Domino folder run `docker-compose -f docker-compose-dev.yaml`
3. Go to this [file lines](https://github.com/Tauffer-Consulting/domino/blob/61c044360fc20bc9c499ab449ca527527be160fa/src/domino/custom_operators/docker_operator.py#L58-L68) and replace the following:
```
# 1. Activate general hot reloading
dev_pieces = False  # Replace with dev_pieces = True

-------------------
# 2. Activating hot reloading for pieces repositories

local_repos_path = f"/home/vinicius/Documents/work/tauffer/{piece_repo_name}" 

## Replace the above with the path where your pieces repositories are. 
## For example f"/home/tdari/Documents/domino-pieces/{piece_repo_name}" 

-------------------

# 3. Activate hot reloading for local domino-py package
Mount(source='/home/vinicius/Documents/work/tauffer/domino/src/domino', target='/usr/local/lib/python3.10/site-packages/domino/', type='bind', read_only=True) 

# Replace the source path with the path where your local domino-py is. 
# Example Mount(source='/home/tdari/Documents/domin/src/domino', ...)

```

The 1. will activate the mounts when starting each piece.
The 2. will mount a volume when your piece container starts pointing to the local path of your piece repository, overwriting the code of the container with the local code.
The 3. will mount a volume when your piece container starts poiting to the path where domino-py is instaled in the container, overwriting the code of domino-py in the container with your local domino-py.


Also we have an in progress PR for the docs for this kind of stuff, you can follow it [here](https://github.com/Tauffer-Consulting/domino-docs/pull/32).


Please try the approach I described until we have the interface. Let me know if need more help.
Thanks!

**Reply by tdari** [reply]
Thank you, this is working, you saved me :) thanks, guys @vinicvaz @luiztauffer
